File: src/com/mishiranu/dashchan/content/net/GithubRepository.java
package com.mishiranu.dashchan.content.net;

import android.net.Uri;
import chan.content.Chan;
import chan.http.HttpException;
import chan.http.HttpHolder;
import chan.http.HttpRequest;
import java.util.LinkedHashMap;
import java.util.Map;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

/** Lists and downloads repository files through the GitHub REST API. */
public class GithubRepository {
	public static class Entry {
		public final boolean directory;
		public final Uri downloadUri;

		Entry(boolean directory, Uri downloadUri) {
			this.directory = directory;
			this.downloadUri = downloadUri;
		}
	}

	private final HttpHolder holder;
	private final String repositoryPath;

	/** @param githubUri repository URI, e.g. {@code https://github.com/owner/repo} */
	public GithubRepository(HttpHolder holder, Uri githubUri) {
		this.holder = holder;
		String path = githubUri.getPath();
		if (path == null || path.isEmpty()) {
			throw new IllegalArgumentException("Invalid GitHub URI: " + githubUri);
		}
		repositoryPath = path.endsWith("/") ? path.substring(0, path.length() - 1) : path;
	}

	/** Returns a map of file names to entries for the given repository directory. */
	public Map<String, Entry> listFiles(String path) throws HttpException, JSONException {
		Uri uri = Uri.parse("https://api.github.com").buildUpon()
				.appendEncodedPath("repos" + repositoryPath + "/contents/" + path).build();
		String response = new HttpRequest(uri, holder)
				.addHeader("Accept", "application/vnd.github+json")
				.perform().readString();
		JSONArray array = new JSONArray(response);
		LinkedHashMap<String, Entry> files = new LinkedHashMap<>();
		for (int i = 0; i < array.length(); i++) {
			JSONObject item = array.getJSONObject(i);
			String name = item.getString("name");
			boolean directory = "dir".equals(item.optString("type"));
			String downloadUrl = item.isNull("download_url") ? null : item.optString("download_url", null);
			files.put(name, new Entry(directory, downloadUrl != null ? Uri.parse(downloadUrl) : null));
		}
		return files;
	}

	public byte[] readFile(Entry entry) throws HttpException {
		if (entry.downloadUri == null) {
			return null;
		}
		return new HttpRequest(entry.downloadUri, holder).perform().readBytes();
	}

	public static HttpHolder createHolder() {
		return new HttpHolder(Chan.getFallback());
	}
}

File: src/com/mishiranu/dashchan/content/net/GithubRepository.kt
package com.mishiranu.dashchan.content.net

import android.net.Uri
import chan.content.Chan
import chan.http.HttpException
import chan.http.HttpHolder
import chan.http.HttpRequest
import org.json.JSONArray
import org.json.JSONException

/**
 * Lists and downloads repository files through the GitHub REST API.
 *
 * @param githubUri repository URI, e.g. `https://github.com/owner/repo`
 */
class GithubRepository(
    private val holder: HttpHolder,
    githubUri: Uri,
) {
    class Entry internal constructor(
        @JvmField val directory: Boolean,
        @JvmField val downloadUri: Uri?,
    )

    private val repositoryPath: String

    init {
        val path = githubUri.path
        require(!path.isNullOrEmpty()) { "Invalid GitHub URI: $githubUri" }
        repositoryPath = if (path.endsWith("/")) path.substring(0, path.length - 1) else path
    }

    /** Returns a map of file names to entries for the given repository directory. */
    @Throws(HttpException::class, JSONException::class)
    fun listFiles(path: String): Map<String, Entry> {
        val uri =
            Uri
                .parse("https://api.github.com")
                .buildUpon()
                .appendEncodedPath("repos$repositoryPath/contents/$path")
                .build()
        val response =
            HttpRequest(uri, holder)
                .addHeader("Accept", "application/vnd.github+json")
                .perform()!!
                .readString()
        val array = JSONArray(response)
        val files = LinkedHashMap<String, Entry>()
        for (i in 0 until array.length()) {
            val item = array.getJSONObject(i)
            val name = item.getString("name")
            val directory = "dir" == item.optString("type")
            // isNull() is already true for both an absent key and a JSON null, so in the
            // else branch the value exists and optString() can never fall back to null.
            val downloadUrl = if (item.isNull("download_url")) null else item.optString("download_url")
            files[name] = Entry(directory, if (downloadUrl != null) Uri.parse(downloadUrl) else null)
        }
        return files
    }

    @Throws(HttpException::class)
    fun readFile(entry: Entry): ByteArray? {
        val downloadUri = entry.downloadUri ?: return null
        return HttpRequest(downloadUri, holder).perform()!!.readBytes()
    }

    companion object {
        @JvmStatic
        fun createHolder(): HttpHolder = HttpHolder(Chan.getFallback())
    }
}
